### ISSUE TITLE
Fine-tuning SuperPoint

### ISSUE THREAD

**Issue by MatteoCappe** [MatteoCappe/Matteo_thesis_workbook#31]
Github repository: https://github.com/eric-yyjau/pytorch-superpoint

## Dataset creation

DSEC recitifed images: 

**Training**: thun_a_00 and zurich_city_11_c
**Validation**: interalken_00_c

## Losses:

**Sparse**: optimization attuated by the authors of the pytorch implementation
**Dense**: original loss used to train descriptors from the SuperPoint paper

Known problem: workstation has limited amount of GPU -> can't run training with **dense loss** with batch size 32 or 16, 8 or 10 works best

## Additions

function `visualize_validation_keypoints` inside the file `Train_model_frontend.py` to visualize the keypoints overlayed on images as the training proceeds
function `extract_keypoints_and_descriptors` inside the file `export.py`, to extract both keypoints and descriptor with a single command
file `DSEC.py` in order to run the training on a the DSEC dataset
weights&biases integration to log the loss as graph

## Running the training

Training can be ran either from scratch or starting from a pretrained model/checkpoint
In order to start from a pretrained model/checkpoint just integrate the following line at the end of the config_file.yaml you are using
Also reference the file `training_tutorial.md`: https://github.com/eric-yyjau/pytorch-superpoint/blob/4ff74df8fa3c10ce9eb9fdc561f787d9e8bc9691/train_tutorial.md

Follow also the instruction on the ReadMe the author provided: https://github.com/eric-yyjau/pytorch-superpoint/blob/4ff74df8fa3c10ce9eb9fdc561f787d9e8bc9691/README.md

Step 1) Extract pseudo-GroundTruth, both on train and val data (change the field "export_folder" in the config file), with the following command 

```bash
python export.py export_detector_homoAdapt configs/magicpoint_dsec_export.yaml output_folder
```

Step 2) Run the training, remember to check the config files (sorry if they are messy)

```bash
python train4.py train_joint configs/superpoint_dsec_train_heatmap.yaml checkpoints_output_folder
```

### Run on COCO dataset

i don't have any results from mid-training
Ran for 100000 iterations (don't remember how many epochs sorry)
Ran both with and without data adaptation, it follows the same behaviour either way

![Image](https://github.com/user-attachments/assets/e611bc36-c7f9-4ab5-9063-9436fd73c7f6)

### Runs on pretrained weights on DSEC dataset

These are the weights provided from the authors after training on COCO
Ran for 12000 iterations, about 250 epochs

![Image](https://github.com/user-attachments/assets/3b09610c-1833-4e98-bf26-c7dbca332c8b)

Mid-training Results:

Epoch 1:
![Image](https://github.com/user-attachments/assets/3a0007e7-113c-4bcb-a264-5028be0852a6)

Epoch 50_

![Image](https://github.com/user-attachments/assets/fc562e39-fa73-4d3f-80d2-58525df62439)

Epoch 100:

![Image](https://github.com/user-attachments/assets/c087b4ea-b23a-4cce-a5c6-df749f804523)

Critical Points, epochs from 125 to 135

![Image](https://github.com/user-attachments/assets/42548524-7f70-48fc-a097-28d0bfe68089)

![Image](https://github.com/user-attachments/assets/d66caa3d-f82b-4d4b-8ab8-6f4cd30f22ea)

![Image](https://github.com/user-attachments/assets/14d92264-54cd-414e-b1d4-2c7fe24508f1)

![Image](https://github.com/user-attachments/assets/43ab6827-2e87-4d58-9d6f-c694d5fc84dd)

![Image](https://github.com/user-attachments/assets/84f99730-e87d-4f40-962b-1500a5b7f2f7)

![Image](https://github.com/user-attachments/assets/acfd4811-f9f9-4dbe-b6e6-e2d2b316d2b0)

![Image](https://github.com/user-attachments/assets/0a2f08a7-ecf5-49d4-9064-d9d4e70c8b3f)

![Image](https://github.com/user-attachments/assets/ff57a0e0-4fd2-4d2b-b85f-8dfa3713ae9c)

![Image](https://github.com/user-attachments/assets/c495f6b3-53b4-443b-ab82-55e6befd7408)

![Image](https://github.com/user-attachments/assets/31588f38-7507-4d8b-89d2-37284af17f45)

![Image](https://github.com/user-attachments/assets/d5911b58-99ab-4ab3-97da-14bbd4e44be0)

Epoch 150: 

![Image](https://github.com/user-attachments/assets/511fb1c9-cc91-4e31-b700-81f09882e7b6)

Epoch 200 (Validation):

![Image](https://github.com/user-attachments/assets/4df3c2e8-2e3f-4990-9d96-3061578b7050)

### Runs from scratch on DSEC dataset

Validation every 200 epochs, I'm only showing the validation steps

Ran twice, one with batch size 8 and dense loss and the other with batch size 32 and sparse loss

### Results from Batch_size 8 and dense loss

Ran for 1000 epochs

![Image](https://github.com/user-attachments/assets/640e4c52-2f6c-4bc3-9de7-c7e7afed7055)

Note: i'm not sure if the validation epochs on the x axis are correct i may have logged them wrongly

Epoch 1:

![Image](https://github.com/user-attachments/assets/938bcf2a-a030-48ad-9373-c93b198d7955)

Epoch 200:

![Image](https://github.com/user-attachments/assets/1a679140-c2a5-452f-a58d-6a110905b0c6)

Epoch 400: 

![Image](https://github.com/user-attachments/assets/58f8ef1b-eda5-4ecc-b3b6-5406b1336f64)

Epoch 600:

![Image](https://github.com/user-attachments/assets/e62a02db-eb40-4768-a931-cab739302a03)

Epoch 800:

![Image](https://github.com/user-attachments/assets/7339e113-9264-45e6-8e5f-6035685506e8)

Epoch 1000:

![Image](https://github.com/user-attachments/assets/e55fd9ec-d922-4404-80ca-3888422bec52)

### Results from Batch_size 32 and sparse loss

Ran for 1800 epochs

![Image](https://github.com/user-attachments/assets/c4ef76d2-6485-420a-8e00-ef39f319ddbc)

Epoch 1:

![Image](https://github.com/user-attachments/assets/938bcf2a-a030-48ad-9373-c93b198d7955)

Epoch 200:

![Image](https://github.com/user-attachments/assets/37187d9a-d1af-4f42-b4db-50eebfc49864)

Epoch 400:

![Image](https://github.com/user-attachments/assets/fd5b456e-c7a1-4c89-a0b0-7e537db71fd8)

Epoch 600: 

![Image](https://github.com/user-attachments/assets/a50d2e90-8e13-441f-9353-39ecf50084ee)

Epoch 800:

![Image](https://github.com/user-attachments/assets/2bb2dff6-0cf7-417f-b267-22691d3db3b4)

Epoch 1000:

![Image](https://github.com/user-attachments/assets/bcd9a1fe-b8a2-4fe7-bd8d-52cb378b47ee)

Epoch 1200:

![Image](https://github.com/user-attachments/assets/f4b7b7b5-1ca6-42c8-ba63-b924cd574e8b)

Epoch 1400:

![Image](https://github.com/user-attachments/assets/0623dd54-576e-40e4-b1c3-a4d79745b153)

Epoch 1600:

![Image](https://github.com/user-attachments/assets/f5ba3fad-8aac-435f-94aa-40a49cf5be62)

Epoch 1800:

![Image](https://github.com/user-attachments/assets/5a0d01c6-4fa3-4bc2-a835-9b39fdd04b4b)

Note: i just realized that i should try with "white representation" instead of this one, maybe something changes idk

## Exporting keypoints and descriptors

The export can be done with the following command

```bash
python export.py extract_keypoints_and_descriptors configs/magicpoint_repeatability_heatmap.yaml output_folder
```

Note: i changed the original evaluation script as it was ran on hpatches and i didn't care about that experiment, see here the original version: https://github.com/eric-yyjau/pytorch-superpoint/blob/master/configs/magicpoint_repeatability_heatmap.yaml

Note: I should check this again

## Evalution

Refer to the following issue to understand how the network behaves with pretrained models: https://github.com/MatteoCappe/Matteo_thesis_workbook/issues/27

Extract keypoints and descriptors adn then run evaluation

The evaluation on the points and descriptors extracted with the pytorch implementation can be ran with the following script: https://github.com/MatteoCappe/Matteo_thesis_workbook/blob/main/DSEC_evaluation_pytorch/DSEC_evaluation_matches_correctness.py

### Weights at 1000th epoch from the trainig from scratch with batch size 8

Left image of thun:

![Image](https://github.com/user-attachments/assets/dd00ecdd-35f9-4670-8f8f-142ccd97d1e9)

corresponding Right image of thun:

![Image](https://github.com/user-attachments/assets/9c5728b3-9cfd-41c0-b227-2c95411adee6)

Matches: 0, accuracy 0%

### Weird things

I tried this with other weights nont the ones just mentioned
I tried matching points on the left frame with points on the left frame (since all points should have been matched with each other) and noticed that it varies a lot during the whole sequence and i never get the amount of total matches equal to the amount of keypoints

![Image](https://github.com/user-attachments/assets/e7afa1fe-e1c0-4da9-9220-b2502d5a0fed)
![Image](https://github.com/user-attachments/assets/66eda76f-0ad5-4bed-b7d2-8a83ed9a8dc9)
